### DISCUSSION TITLE
Extension vs Module vs Profile

### DISCUSSION THREAD

**Discussion by kristojorg** [readium/webpub-manifest#94] [Q&A] [answered]
What is meant to be the functional difference between an extension, module and a profile? A profile seems to declare that a manifest uses some extension, but there doesn't seem to be a way to declare that a manifest uses a module. What is the relationship between these three concepts?

**Comment by kristojorg** [comment, marked as answer]
Furthermore, how does the [presentation hints module](https://github.com/readium/webpub-manifest/blob/master/modules/presentation.md) square with the [epub profile](https://github.com/readium/webpub-manifest/blob/master/profiles/epub.md)? 

1. The epub profile allows for `properties.layout` in a link, which seems to conflict or at least overlap with the `properties.presentation` in presentation hints. 
2. The [epub extension schema](https://github.com/readium/webpub-manifest/blob/master/schema/extensions/epub/metadata.schema.json) (is this the same as the epub profile?) allows for a `presentation.layout` property in the manifest's metadata, which conflicts with the `presentation` property defined in the presentation hints module. Note: I made a PR to document this metadata property in the Epub profile docs [here](https://github.com/readium/webpub-manifest/pull/86)

**Reply by mickael-menu** [reply]
"Extension" is the former name for both modules and profiles (they got split, see https://github.com/readium/webpub-manifest/issues/46). The schema are still stored under `extensions/` probably to avoid breaking links.

A profile it a set of mutually exclusive extensions (in the sense of JSON extensibility) to RWPM to support a particular kind of publication. A manifest needs to declare which profile it implements because this is often the first step in the apps to determine what to do with the publication (e.g. which navigator to use).

A module is also a JSON extension to RWPM, but can be applied to multiple profiles. Apps can just check for their presence in a manifest on the fly. There's only one real module at the moment: **Encryption**, which can be used with all profiles.

Regarding the Presentation Hints module, it will probably get removed (or renamed into Divina Presentation Hints), see https://github.com/readium/webpub-manifest/pull/89. In which case it might just be part of the Divina profile and not be a module anymore.

**Reply by kristojorg** [reply]
Okay this is great information, thank you. If I'm understanding this correctly, and a module is essentially an optional property, why distinguish it from any other optional property?

**Reply by mickael-menu** [reply]
The models evolved organically. In the case of `encrypted`, [it used to be simply an optional property of the EPUB profile](https://github.com/readium/webpub-manifest/issues/42).

But in my understanding, RWPM contains core properties (potentially optional) describing the publication. Modules enable additional functionalities that relate only tangentially to the publication itself. That's why IMHO the Presentation Hints should not be a module, as it's core to the representation of a Divina.

**Reply by kristojorg** [reply]
I see, thank you. In my opinion, the modules distinction is a bit unnecessary or unclear. For example, I'm not sure why the `duration` or `bitrate` properties would be an optional link property while `encryption` is a separate module. In this case the `encryption` property actually seems to be more generic. Any resource could be encrypted while only audiobook resources could have a `duration`. 

My personal recommendation would be to do away with the module concept as a whole and simply have the link properties contain any optional properties that are profile-agnostic and any profile-specific properties be included in the profile (divina presentation hints for example)

**Reply by kristojorg** [reply]
That being said, I'm obviously missing a lot of context and y'all seem to have good reasons for things in my experience : ). Just wanted to toss in my 2 cents in case it's useful

**Reply by mickael-menu** [reply]
And we might do just that when Encryption is the only module left. Although I can see why functionally Encryption is not a core property and we might want to keep it separate.

Moving `duration` and `bitrate` to the Audiobook and Divina profiles could be a solution too. I found these properties a bit weird in the core spec.

Thank you, an external look is always useful.

**Reply by HadrienGardeur** [reply]
I don't think that moving `duration` or `bitrate` to the Audiobook or Divina profiles would be a good idea.

They both apply to any audio or video resource, which is generic enough to warrant a presence as a core member of the Link Object. The same can be said about `height` and `width` which both apply to any image or video resource.

**Reply by kristojorg** [reply]
I think that's fine @HadrienGardeur , but would recommend either clarifying what a "module" is, or simply doing away with the concept altogether. It seem the Link Object could have `duration`, `bitrate`, `height`, `width` and `encryption` all as optional properties. I'm not entirely sure what makes `encryption` a "module" here while the others are currently optional properties. Personally, I would do away with the concept of modules altogether and leave just optional properties and profiles.
